engine/minskyTensorOps.cc

```diff
@@ -446,21 +449,45 @@ namespace minsky
       return nan("");
     }
   };
+   
+  // Original code
+
+//  class RavelTensor: public civita::CachedTensorOp
+//  {
+//    const Ravel& ravel;
+//    TensorPtr arg;
+//    void computeTensor() const override  
+//    {
+//      const_cast<Ravel&>(ravel).loadDataCubeFromVariable(*arg);      
+//      ravel.loadDataFromSlice(cachedResult);
+//      m_timestamp = Timestamp::clock::now();
+//    }    
+//    CLASSDESC_ACCESS(Ravel);
+//  public:
+//    RavelTensor(const Ravel& ravel): ravel(ravel) {}
+//    void setArgument(const TensorPtr& a,const std::string&,double) override {arg=a;
+//  	; cachedResult.index(a->index()); cachedResult.hypercube(a->hypercube());}
+//    Timestamp timestamp() const override {return arg? arg->timestamp(): Timestamp();}
+//  };
 
-  class RavelTensor: public CachedTensorOp
+// Alternative version of original code with ravel op added in tensorOpFactory
+  class RavelTensor: public civita::CachedTensorOp
   {
-    const Ravel& ravel;
+    CLASSDESC_ACCESS(Ravel);	  
     TensorPtr arg;
-    void computeTensor() const override
+    void computeTensor() const override  
     {
-      const_cast<Ravel&>(ravel).loadDataCubeFromVariable(*arg);
-      ravel.loadDataFromSlice(cachedResult);
-      m_timestamp = Timestamp::clock::now();
-    }
-    
+	   if (auto r=dynamic_cast<Ravel*>(arg.get())) {
```

Comment:
This sounds like a confusion. arg points to an ITensor object, Ravel is an Item. Two different object heirarchies. Maybe your thinking of subclassing Ravel from ITensor, but I'd rather you didn't. I don't think it would make sense.

Reply:
Confusion cleared up. Registering RavelTensor as an Op in tensorFactoryOp like TimeOp caused problems with the Ravel object declared as a reference in the RavelTensor class. I got the uninitialised reference error.  So I made "const Ravel& ravel" just plain "const Ravel ravel". 

Of course, then, I didn't get the updated Ravel state...I tried making it a shared_ptr and also a unique_ptr, but Minsky would crash even just for simple tensors like iota(9)... Finally, it was also necessary to add the default constructor of RavelTensor in order to compile the code.

So, I gave up and in the end moved the Item-to-Ravel cast to the top of the if conditional in TensorOpFactory::create() and just updated cachedResult in setArguments in RavelTensor, which has got me the closest to version 2.15.5 functionality I've seen.

Now, cachedResult will have to be updated by a chain of operations in RavelTensor::ComputeTensor() if it is to work properly. 

But I'm not yet fully convinced I can't take advantage of what's already in ravelWrap. Are any of the other members in ravelWrap available to me, e.g., the slicing and sorting?

In this regard, I'll soon be committing code that uses the C interface in ravelWrap.cc in the hope that cachedResult is updated correctly. To use it with the tensor library wouldn't I just have to use the correct hypercube indices? At present, it compiles but then crashes when I try to test it on iota(9) or similar...

Reply:
It took duplicating within minskyTensorOps.cc a lot of code from ravelWrap.cc to figure out that the last line in Ravel::loadDataFromSlice() was returning an empty result in the original code. Now, if I understand correctly, I have to replace the "ravel_" functions in Ravel::loadDataCubeFromVariable() and Ravel::loadDataFromSlice() with versions that are consistent with the Civita library. This is what I've started doing today.